### ISSUE TITLE
includeResultMetadata types not changing return type?

### ISSUE THREAD

**Issue by simllll** [Automattic/mongoose#14303]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported

### Mongoose version

8.1.1

### Node.js version

20.x

### MongoDB server version

6.x

### Typescript version (if applicable)

5.1.6

### Description

I tried upgrading to mongoose 8.x and changed rawResult: true to includeResultMetadata: true, but now typescript complains:
```
TS2339: Property value does not exist on type
FlattenMaps<IDBJobField> & Required<{   _id: ObjectId; }>

```

also it seems that the retutrn type is not changed depending on the value of includeResultMetadata
![image](https://github.com/Automattic/mongoose/assets/5757263/1c397a9e-b06a-4a3d-8a52-90d76227c79c)
it's the same as in the screenshot, regardless of this flag.

Furthermore; i found it kinda confusing with the mongodb docs, as they say that includeResultMetadata is true by default. Does mongoose set it to false then by default? Or is it actually true, but mongoose just returns the document and strips the rest away? Which would sound to me of a waste of some extra bytes?

error TS2339: Property 'value' does not exist on type 'FlattenMaps<IDBJobField> & Required<{ _id: ObjectId; }>'.
error TS2339: Property 'lastErrorObject' does not exist on type 'FlattenMaps<IDBJobField> & Required<{ _id: ObjectId; }>'.
error TS2339: Property 'value' does not exist on type 'FlattenMaps<IDBJobField> & Required<{ _id: ObjectId; }>'.

### Steps to Reproduce

use an example query for findOneAndupdate and play around with the includeResultMetadata flag

### Expected Behavior

should return the "ModifyResult" instead, like it did with rawResult: true

**Comment by vkarpov15** [comment]
Where in the docs do you see that `includeResultMetadata` is true by default? I think that changed in MongoDB Node driver 6.

Also, is the error just in your IDE or also in TypeScript? Because I'm unable to repro with TypeScript, the following script compiles fine:

```ts
import mongoose from 'mongoose';

run();

async function run() {
  const TestModel = mongoose.model('Test', new mongoose.Schema({ name: String }));
  const doc = await TestModel.findOneAndUpdate({}, { name: 'bar' }, { includeResultMetadata: true });
  doc.value;
}
```

```
$ ./node_modules/.bin/tsc --strict gh-14303.ts 
$ 
```

**Comment by simllll** [comment]
Regarding includeResultMetadata,
you are right, it's mentionted here that it will be set to false. https://github.com/mongodb/node-mongodb-native/releases/tag/v6.0.0

> Once 6.0.0 is released, includeResultMetadata: false will become the default behavior. If your application relies on the previous behavior of these APIs, setting includeResultMetadata: true will allow you to continue to access the ModifyResult directly.

![image](https://github.com/Automattic/mongoose/assets/5757263/b0b014e7-fa18-4343-9591-99d7beafbac9)

so this should be safe, it wsa just kinda of confusing ;-)

regarding the typescript issue, I iwll look into this now

**Comment by simllll** [comment]
Alright, I found the issue. It's about the "lean" flag.

```ts
import mongoose from 'mongoose';

run();

async function run() {
  const TestModel = mongoose.model('Test', new mongoose.Schema({ name: String }));
  const doc = await TestModel.findOneAndUpdate({}, { name: 'bar' }, { includeResultMetadata: true, lean: true, returnDocument: 'after' });
  doc.value;
}
```

the problem is when I add "lean: true", which brings me to one followup question: If we set includeResultMetadata to true, is mongoose then populating stuff and returns a mongoose document, or does it actually return the raw result like with lean: true?
Because the type of includeResultMetadta: false is the exact same as the ".value" prop when I set it to true...

**Comment by vkarpov15** [comment]
**This brings me to another followup question: If we set includeResultMetadata to true, is mongoose then populating stuff and returns a mongoose document, or does it actually return the raw result like with lean: true?** If `lean` is false, then Mongoose will fully hydrate the document, and `.value` will contain the fully hydrated document.

**I would expect it to be the lean version, as this was the behaviour with "rawResult" before. But to be explicit about that, what happens then if I set lean explicity to false?** Setting `lean: false` is the same as not setting `lean` at all in this case.
